File: backend/src/main/java/com/smartpark/dto/response/ApiResponse.java
// Standard API response wrapper for all endpoints.
package com.smartpark.dto.response;

public class ApiResponse<T> {

    private boolean success;
    private T data;
    private String message;
    private ErrorDetails error;

    private ApiResponse() {}

    public static <T> ApiResponse<T> success(T data) {
        ApiResponse<T> response = new ApiResponse<>();
        response.success = true;
        response.data = data;
        return response;
    }

    public static <T> ApiResponse<T> success(T data, String message) {
        ApiResponse<T> response = new ApiResponse<>();
        response.success = true;
        response.data = data;
        response.message = message;
        return response;
    }

    public static <T> ApiResponse<T> error(String code, String message) {
        ApiResponse<T> response = new ApiResponse<>();
        response.success = false;
        response.error = new ErrorDetails(code, message);
        return response;
    }

    public boolean isSuccess() { return success; }
    public T getData() { return data; }
    public String getMessage() { return message; }
    public ErrorDetails getError() { return error; }

    public static class ErrorDetails {
        private final String code;
        private final String message;

        public ErrorDetails(String code, String message) {
            this.code = code;
            this.message = message;
        }

        public String getCode() { return code; }
        public String getMessage() { return message; }
    }
}

File: frontend/lib/models/api_response.dart
// api_response.dart — Generic API response wrapper matching backend ApiResponse<T>
import 'package:flutter/foundation.dart';

/// Represents the standard error payload returned by the backend.
@immutable
class ApiError {
  /// Machine-readable error code (e.g. 'NOT_FOUND', 'BADGE_SUSPENDED').
  final String code;

  /// Human-readable error message shown in the UI.
  final String message;

  /// Creates an [ApiError].
  const ApiError({required this.code, required this.message});

  /// Deserialises from the backend `error` object.
  factory ApiError.fromJson(Map<String, dynamic> json) => ApiError(
        code: json['code'] as String,
        message: json['message'] as String,
      );
}

/// Generic wrapper for every backend response envelope.
///
/// Use [fromJson] with a [fromData] converter to parse [T].
@immutable
class ApiResponse<T> {
  /// Whether the request succeeded.
  final bool success;

  /// Parsed response payload; null on error.
  final T? data;

  /// Optional human-readable success message.
  final String? message;

  /// Error detail; null on success.
  final ApiError? error;

  /// Creates an [ApiResponse].
  const ApiResponse({
    required this.success,
    this.data,
    this.message,
    this.error,
  });

  /// Deserialises the JSON envelope and uses [fromData] to parse [T].
  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic)? fromData,
  ) {
    final errorJson = json['error'] as Map<String, dynamic>?;
    final rawData = json['data'];

    return ApiResponse<T>(
      success: json['success'] as bool,
      data: (fromData != null && rawData != null) ? fromData(rawData) : null,
      message: json['message'] as String?,
      error: errorJson != null ? ApiError.fromJson(errorJson) : null,
    );
  }
}
